llvm/lib/Transforms/InstCombine/InstCombineLoadStoreAlloca.cpp

```diff
@@ -1065,6 +1065,25 @@ Instruction *InstCombinerImpl::visitLoadInst(LoadInst &LI) {
         V1->setAtomic(LI.getOrdering(), LI.getSyncScopeID());
         V2->setAlignment(Alignment);
         V2->setAtomic(LI.getOrdering(), LI.getSyncScopeID());
+        // Can copy any metadata that cannot trigger UB. In particular do not
+        // copy !noundef or !invariant.load.
+        unsigned SafeMDKinds[] = {LLVMContext::MD_dbg,
+                                  LLVMContext::MD_tbaa,
+                                  LLVMContext::MD_prof,
+                                  LLVMContext::MD_fpmath,
+                                  LLVMContext::MD_tbaa_struct,
+                                  LLVMContext::MD_alias_scope,
+                                  LLVMContext::MD_noalias,
+                                  LLVMContext::MD_nontemporal,
+                                  LLVMContext::MD_mem_parallel_loop_access,
+                                  LLVMContext::MD_access_group,
```

Comment:
It's my understanding that we currently assume all AA-related metadata to be IUB. It's possible that we can relax some of it to return poison for loads and only be IUB, but if we want to do that, I'd expect this to start with a LangRef change. We should also keep https://github.com/llvm/llvm-project/blob/ccaded2b1d0d2cf3d8041baeeec9cfad632c9450/llvm/lib/IR/Instruction.cpp#L458-L468 in sync if we do that.

Reply:
Ah thanks, this is something that wasn't really clear to me when reading the langref. Using the fact that they are UB, we should be able to retain the metadata if the instruction doesn't move, put up https://github.com/llvm/llvm-project/pull/115868

Reply:
(I'll update the patch to limit it to the kinds marked as `hasPoisonGeneratingMetadata` for now)